modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -512,11 +564,24 @@ def map_func(left, right=right, kwargs=kwargs):
     def reindex(self, axis, labels, **kwargs):
         new_index = self.index if axis else labels
         new_columns = labels if axis else self.columns
+        actual_index = None
+        if "_reset_index" in kwargs:
+            actual_index = kwargs.pop("_reset_index")
+
+        def _reindex(df):
+            df = df.reindex(labels=labels, axis=axis, **kwargs)
+            if actual_index is not None:
+                df.index = actual_index
+            return df
+
```

Comment:
Reindexing for operation `join` needs to be done in a special way. First, reindexing is performed by the index obtained from the names in `on` variable, and then it needs to immediately update the internal index of right operand to the one used for the left operand (this will help to avoid expensive index synchronization)